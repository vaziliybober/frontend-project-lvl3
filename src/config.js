export default {
  updateInterval: 5000,
  proxyUrl: 'https://cors-anywhere.herokuapp.com',
};
