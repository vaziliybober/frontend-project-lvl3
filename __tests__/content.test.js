import nock from 'nock';
import prettier from 'prettier';
import path from 'path';
import fs, { promises as fsp } from 'fs';
import { waitFor } from '@testing-library/dom';
import '@testing-library/jest-dom';
import axios from 'axios';
import httpAdapter from 'axios/lib/adapters/http';
import run from '../src/application.js';
import config from '../src/config.js';

axios.defaults.adapter = httpAdapter;
nock.disableNetConnect();

const getFixturePath = (filepath) => path.join(__dirname, '..', '__fixtures__', filepath);

const options = {
  parser: 'html',
  htmlWhitespaceSensitivity: 'ignore',
  tabWidth: 4,
};

const { proxyUrl } = config;
const someUrl = 'https://ru.hexlet.io/lessons.rss';

let elements;

const prettify = (html) => prettier.format(html, options);
const getSectionTree = () => prettify(elements.section.innerHTML);

beforeEach(() => {
  const initHtml = fs.readFileSync(path.join(__dirname, '..', 'index.html')).toString();
  document.documentElement.innerHTML = initHtml;

  const section = document.querySelector('.section-content');
  const form = document.querySelector('form');

  elements = {
    section,
    input: form.querySelector('input'),
    submit: form.querySelector('button'),
    feedsContainer: section.querySelector('.feeds'),
    postsContainer: section.querySelector('.posts'),
  };
});

test('content initial', async () => {
  run();
  expect(getSectionTree()).toMatchSnapshot();
});

test('content feeds with updating', async () => {
  const oldSetTimeout = window.setTimeout;
  window.setTimeout = (callback) => oldSetTimeout(callback, 10);

  run();

  const rss = await fsp.readFile(getFixturePath('rss.xml'));
  const rss1 = await fsp.readFile(getFixturePath('rss1.xml'));
  const rss2 = await fsp.readFile(getFixturePath('rss2.xml'));

  elements.input.value = someUrl;
  nock(proxyUrl).get(`/${someUrl}`)
    .reply(200, rss);
  elements.submit.click();
  await waitFor(() => {
    expect(elements.section).toHaveTextContent('Новые уроки на Хекслете');
  });
  expect(getSectionTree()).toMatchSnapshot();

  elements.input.value = someUrl;
  nock(proxyUrl).get(`/${someUrl}`)
    .reply(200, rss1);
  elements.submit.click();
  await waitFor(() => {
    expect(elements.section).toHaveTextContent('OLD ITEM');
    expect(elements.section).not.toHaveTextContent('NEW ITEM');
  });
  expect(getSectionTree()).toMatchSnapshot();

  const scope = nock(proxyUrl).get(`/${someUrl}`)
    .reply(200, rss2);
  await waitFor(() => {
    expect(elements.section).toHaveTextContent('NEW ITEM');
  });
  expect(getSectionTree()).toMatchSnapshot();
  scope.done();

  window.setTimeout = oldSetTimeout;
});
