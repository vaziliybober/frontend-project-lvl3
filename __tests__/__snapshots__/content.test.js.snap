// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`content feeds with updating 1`] = `
"<div class=\\"row mb-5\\">
    <div class=\\"feeds col-md-10 col-lg-8 mx-auto\\">
        <h2>Feeds</h2>
        <ul class=\\"list-group\\">
            <li class=\\"list-group-item\\">Новые уроки на Хекслете</li>
        </ul>
    </div>
</div>
<div class=\\"row\\">
    <div class=\\"posts col-md-10 col-lg-8 mx-auto\\">
        <h2>Posts</h2>
        <ul class=\\"list-group\\">
            <li class=\\"list-group-item\\">
                <a
                    href=\\"https://ru.hexlet.io/ ourses/js-object-oriented-design/lessons/proxy/theory_unit\\"
                    target=\\"_blank\\"
                    rel=\\"nofollow\\"
                >
                    Прокси (Proxy) / JS: Объектно-ориентированный дизайн
                </a>
            </li>
            <li class=\\"list-group-item\\">
                <a
                    href=\\"https://ru.hexlet.io/courses/js-object-oriented-design/lessons/right-way-to-write-code/theory_unit\\"
                    target=\\"_blank\\"
                    rel=\\"nofollow\\"
                >
                    Пишем код правильно / JS: Объектно-ориентированный дизайн
                </a>
            </li>
            <li class=\\"list-group-item\\">
                <a
                    href=\\"https://ru.hexlet.io/courses/js-object-oriented-design/lessons/modeling/theory_unit\\"
                    target=\\"_blank\\"
                    rel=\\"nofollow\\"
                >
                    Объекты-Сущности, Объекты-Значения и внедренные объекты /
                    JS: Объектно-ориентированный дизайн
                </a>
            </li>
            <li class=\\"list-group-item\\">
                <a
                    href=\\"https://ru.hexlet.io/courses/js-classes/lessons/about/theory_unit\\"
                    target=\\"_blank\\"
                    rel=\\"nofollow\\"
                >
                    О курсе / JS: Погружаясь в классы
                </a>
            </li>
            <li class=\\"list-group-item\\">
                <a
                    href=\\"https://ru.hexlet.io/courses/js-classes/lessons/inheritance/theory_unit\\"
                    target=\\"_blank\\"
                    rel=\\"nofollow\\"
                >
                    Наследование / JS: Погружаясь в классы
                </a>
            </li>
            <li class=\\"list-group-item\\">
                <a
                    href=\\"https://ru.hexlet.io/courses/js-object-oriented-design/lessons/fluent-interface/theory_unit\\"
                    target=\\"_blank\\"
                    rel=\\"nofollow\\"
                >
                    Fluent Interface / JS: Объектно-ориентированный дизайн
                </a>
            </li>
            <li class=\\"list-group-item\\">
                <a
                    href=\\"https://ru.hexlet.io/courses/js-object-oriented-design/lessons/patterns/theory_unit\\"
                    target=\\"_blank\\"
                    rel=\\"nofollow\\"
                >
                    Шаблоны Проектирования / JS: Объектно-ориентированный дизайн
                </a>
            </li>
            <li class=\\"list-group-item\\">
                <a
                    href=\\"https://ru.hexlet.io/courses/js-object-oriented-design/lessons/configuration-setters/theory_unit\\"
                    target=\\"_blank\\"
                    rel=\\"nofollow\\"
                >
                    Изменяемая конфигурация / JS: Объектно-ориентированный
                    дизайн
                </a>
            </li>
            <li class=\\"list-group-item\\">
                <a
                    href=\\"https://ru.hexlet.io/courses/js-object-oriented-design/lessons/configuration/theory_unit\\"
                    target=\\"_blank\\"
                    rel=\\"nofollow\\"
                >
                    Конфигурация / JS: Объектно-ориентированный дизайн
                </a>
            </li>
            <li class=\\"list-group-item\\">
                <a
                    href=\\"https://ru.hexlet.io/courses/js-object-oriented-design/lessons/about/theory_unit\\"
                    target=\\"_blank\\"
                    rel=\\"nofollow\\"
                >
                    О курсе / JS: Объектно-ориентированный дизайн
                </a>
            </li>
        </ul>
    </div>
</div>
"
`;

exports[`content feeds with updating 2`] = `
"<div class=\\"row mb-5\\">
    <div class=\\"feeds col-md-10 col-lg-8 mx-auto\\">
        <h2>Feeds</h2>
        <ul class=\\"list-group\\">
            <li class=\\"list-group-item\\">Новые уроки на Хекслете</li>
        </ul>
    </div>
</div>
<div class=\\"row\\">
    <div class=\\"posts col-md-10 col-lg-8 mx-auto\\">
        <h2>Posts</h2>
        <ul class=\\"list-group\\">
            <li class=\\"list-group-item\\">
                <a
                    href=\\"https://ru.hexlet.io/ ourses/js-object-oriented-design/lessons/proxy/theory_unit\\"
                    target=\\"_blank\\"
                    rel=\\"nofollow\\"
                >
                    Прокси (Proxy) / JS: Объектно-ориентированный дизайн
                </a>
            </li>
            <li class=\\"list-group-item\\">
                <a
                    href=\\"https://ru.hexlet.io/courses/js-object-oriented-design/lessons/right-way-to-write-code/theory_unit\\"
                    target=\\"_blank\\"
                    rel=\\"nofollow\\"
                >
                    Пишем код правильно / JS: Объектно-ориентированный дизайн
                </a>
            </li>
            <li class=\\"list-group-item\\">
                <a
                    href=\\"https://ru.hexlet.io/courses/js-object-oriented-design/lessons/modeling/theory_unit\\"
                    target=\\"_blank\\"
                    rel=\\"nofollow\\"
                >
                    Объекты-Сущности, Объекты-Значения и внедренные объекты /
                    JS: Объектно-ориентированный дизайн
                </a>
            </li>
            <li class=\\"list-group-item\\">
                <a
                    href=\\"https://ru.hexlet.io/courses/js-classes/lessons/about/theory_unit\\"
                    target=\\"_blank\\"
                    rel=\\"nofollow\\"
                >
                    О курсе / JS: Погружаясь в классы
                </a>
            </li>
            <li class=\\"list-group-item\\">
                <a
                    href=\\"https://ru.hexlet.io/courses/js-classes/lessons/inheritance/theory_unit\\"
                    target=\\"_blank\\"
                    rel=\\"nofollow\\"
                >
                    Наследование / JS: Погружаясь в классы
                </a>
            </li>
            <li class=\\"list-group-item\\">
                <a
                    href=\\"https://ru.hexlet.io/courses/js-object-oriented-design/lessons/fluent-interface/theory_unit\\"
                    target=\\"_blank\\"
                    rel=\\"nofollow\\"
                >
                    Fluent Interface / JS: Объектно-ориентированный дизайн
                </a>
            </li>
            <li class=\\"list-group-item\\">
                <a
                    href=\\"https://ru.hexlet.io/courses/js-object-oriented-design/lessons/patterns/theory_unit\\"
                    target=\\"_blank\\"
                    rel=\\"nofollow\\"
                >
                    Шаблоны Проектирования / JS: Объектно-ориентированный дизайн
                </a>
            </li>
            <li class=\\"list-group-item\\">
                <a
                    href=\\"https://ru.hexlet.io/courses/js-object-oriented-design/lessons/configuration-setters/theory_unit\\"
                    target=\\"_blank\\"
                    rel=\\"nofollow\\"
                >
                    Изменяемая конфигурация / JS: Объектно-ориентированный
                    дизайн
                </a>
            </li>
            <li class=\\"list-group-item\\">
                <a
                    href=\\"https://ru.hexlet.io/courses/js-object-oriented-design/lessons/configuration/theory_unit\\"
                    target=\\"_blank\\"
                    rel=\\"nofollow\\"
                >
                    Конфигурация / JS: Объектно-ориентированный дизайн
                </a>
            </li>
            <li class=\\"list-group-item\\">
                <a
                    href=\\"https://ru.hexlet.io/courses/js-object-oriented-design/lessons/about/theory_unit\\"
                    target=\\"_blank\\"
                    rel=\\"nofollow\\"
                >
                    О курсе / JS: Объектно-ориентированный дизайн
                </a>
            </li>
            <li class=\\"list-group-item\\">
                <a
                    href=\\"http://example.com/test/1604270940\\"
                    target=\\"_blank\\"
                    rel=\\"nofollow\\"
                >
                    Lorem ipsum 2020-11-01T22:49:00Z
                </a>
            </li>
            <li class=\\"list-group-item\\">
                <a
                    href=\\"http://example.com/test/1604270910\\"
                    target=\\"_blank\\"
                    rel=\\"nofollow\\"
                >
                    Lorem ipsum 2020-11-01T22:48:30Z
                </a>
            </li>
            <li class=\\"list-group-item\\">
                <a
                    href=\\"http://example.com/test/1604270880\\"
                    target=\\"_blank\\"
                    rel=\\"nofollow\\"
                >
                    Lorem ipsum 2020-11-01T22:48:00Z
                </a>
            </li>
            <li class=\\"list-group-item\\">
                <a
                    href=\\"http://example.com/test/1604270850\\"
                    target=\\"_blank\\"
                    rel=\\"nofollow\\"
                >
                    Lorem ipsum 2020-11-01T22:47:30Z
                </a>
            </li>
            <li class=\\"list-group-item\\">
                <a
                    href=\\"http://example.com/test/1604270820\\"
                    target=\\"_blank\\"
                    rel=\\"nofollow\\"
                >
                    Lorem ipsum 2020-11-01T22:47:00Z
                </a>
            </li>
            <li class=\\"list-group-item\\">
                <a
                    href=\\"http://example.com/test/1604270790\\"
                    target=\\"_blank\\"
                    rel=\\"nofollow\\"
                >
                    Lorem ipsum 2020-11-01T22:46:30Z
                </a>
            </li>
            <li class=\\"list-group-item\\">
                <a
                    href=\\"http://example.com/test/1604270760\\"
                    target=\\"_blank\\"
                    rel=\\"nofollow\\"
                >
                    Lorem ipsum 2020-11-01T22:46:00Z
                </a>
            </li>
            <li class=\\"list-group-item\\">
                <a
                    href=\\"http://example.com/test/1604270730\\"
                    target=\\"_blank\\"
                    rel=\\"nofollow\\"
                >
                    Lorem ipsum 2020-11-01T22:45:30Z
                </a>
            </li>
            <li class=\\"list-group-item\\">
                <a
                    href=\\"http://example.com/test/1604270700\\"
                    target=\\"_blank\\"
                    rel=\\"nofollow\\"
                >
                    Lorem ipsum 2020-11-01T22:45:00Z
                </a>
            </li>
            <li class=\\"list-group-item\\">
                <a
                    href=\\"http://example.com/test/1604270730\\"
                    target=\\"_blank\\"
                    rel=\\"nofollow\\"
                >
                    OLD ITEM
                </a>
            </li>
        </ul>
    </div>
</div>
"
`;

exports[`content feeds with updating 3`] = `
"<div class=\\"row mb-5\\">
    <div class=\\"feeds col-md-10 col-lg-8 mx-auto\\">
        <h2>Feeds</h2>
        <ul class=\\"list-group\\">
            <li class=\\"list-group-item\\">Новые уроки на Хекслете</li>
        </ul>
    </div>
</div>
<div class=\\"row\\">
    <div class=\\"posts col-md-10 col-lg-8 mx-auto\\">
        <h2>Posts</h2>
        <ul class=\\"list-group\\">
            <li class=\\"list-group-item\\">
                <a
                    href=\\"https://ru.hexlet.io/ ourses/js-object-oriented-design/lessons/proxy/theory_unit\\"
                    target=\\"_blank\\"
                    rel=\\"nofollow\\"
                >
                    Прокси (Proxy) / JS: Объектно-ориентированный дизайн
                </a>
            </li>
            <li class=\\"list-group-item\\">
                <a
                    href=\\"https://ru.hexlet.io/courses/js-object-oriented-design/lessons/right-way-to-write-code/theory_unit\\"
                    target=\\"_blank\\"
                    rel=\\"nofollow\\"
                >
                    Пишем код правильно / JS: Объектно-ориентированный дизайн
                </a>
            </li>
            <li class=\\"list-group-item\\">
                <a
                    href=\\"https://ru.hexlet.io/courses/js-object-oriented-design/lessons/modeling/theory_unit\\"
                    target=\\"_blank\\"
                    rel=\\"nofollow\\"
                >
                    Объекты-Сущности, Объекты-Значения и внедренные объекты /
                    JS: Объектно-ориентированный дизайн
                </a>
            </li>
            <li class=\\"list-group-item\\">
                <a
                    href=\\"https://ru.hexlet.io/courses/js-classes/lessons/about/theory_unit\\"
                    target=\\"_blank\\"
                    rel=\\"nofollow\\"
                >
                    О курсе / JS: Погружаясь в классы
                </a>
            </li>
            <li class=\\"list-group-item\\">
                <a
                    href=\\"https://ru.hexlet.io/courses/js-classes/lessons/inheritance/theory_unit\\"
                    target=\\"_blank\\"
                    rel=\\"nofollow\\"
                >
                    Наследование / JS: Погружаясь в классы
                </a>
            </li>
            <li class=\\"list-group-item\\">
                <a
                    href=\\"https://ru.hexlet.io/courses/js-object-oriented-design/lessons/fluent-interface/theory_unit\\"
                    target=\\"_blank\\"
                    rel=\\"nofollow\\"
                >
                    Fluent Interface / JS: Объектно-ориентированный дизайн
                </a>
            </li>
            <li class=\\"list-group-item\\">
                <a
                    href=\\"https://ru.hexlet.io/courses/js-object-oriented-design/lessons/patterns/theory_unit\\"
                    target=\\"_blank\\"
                    rel=\\"nofollow\\"
                >
                    Шаблоны Проектирования / JS: Объектно-ориентированный дизайн
                </a>
            </li>
            <li class=\\"list-group-item\\">
                <a
                    href=\\"https://ru.hexlet.io/courses/js-object-oriented-design/lessons/configuration-setters/theory_unit\\"
                    target=\\"_blank\\"
                    rel=\\"nofollow\\"
                >
                    Изменяемая конфигурация / JS: Объектно-ориентированный
                    дизайн
                </a>
            </li>
            <li class=\\"list-group-item\\">
                <a
                    href=\\"https://ru.hexlet.io/courses/js-object-oriented-design/lessons/configuration/theory_unit\\"
                    target=\\"_blank\\"
                    rel=\\"nofollow\\"
                >
                    Конфигурация / JS: Объектно-ориентированный дизайн
                </a>
            </li>
            <li class=\\"list-group-item\\">
                <a
                    href=\\"https://ru.hexlet.io/courses/js-object-oriented-design/lessons/about/theory_unit\\"
                    target=\\"_blank\\"
                    rel=\\"nofollow\\"
                >
                    О курсе / JS: Объектно-ориентированный дизайн
                </a>
            </li>
            <li class=\\"list-group-item\\">
                <a
                    href=\\"http://example.com/test/1604270940\\"
                    target=\\"_blank\\"
                    rel=\\"nofollow\\"
                >
                    Lorem ipsum 2020-11-01T22:49:00Z
                </a>
            </li>
            <li class=\\"list-group-item\\">
                <a
                    href=\\"http://example.com/test/1604270910\\"
                    target=\\"_blank\\"
                    rel=\\"nofollow\\"
                >
                    Lorem ipsum 2020-11-01T22:48:30Z
                </a>
            </li>
            <li class=\\"list-group-item\\">
                <a
                    href=\\"http://example.com/test/1604270880\\"
                    target=\\"_blank\\"
                    rel=\\"nofollow\\"
                >
                    Lorem ipsum 2020-11-01T22:48:00Z
                </a>
            </li>
            <li class=\\"list-group-item\\">
                <a
                    href=\\"http://example.com/test/1604270850\\"
                    target=\\"_blank\\"
                    rel=\\"nofollow\\"
                >
                    Lorem ipsum 2020-11-01T22:47:30Z
                </a>
            </li>
            <li class=\\"list-group-item\\">
                <a
                    href=\\"http://example.com/test/1604270820\\"
                    target=\\"_blank\\"
                    rel=\\"nofollow\\"
                >
                    Lorem ipsum 2020-11-01T22:47:00Z
                </a>
            </li>
            <li class=\\"list-group-item\\">
                <a
                    href=\\"http://example.com/test/1604270790\\"
                    target=\\"_blank\\"
                    rel=\\"nofollow\\"
                >
                    Lorem ipsum 2020-11-01T22:46:30Z
                </a>
            </li>
            <li class=\\"list-group-item\\">
                <a
                    href=\\"http://example.com/test/1604270760\\"
                    target=\\"_blank\\"
                    rel=\\"nofollow\\"
                >
                    Lorem ipsum 2020-11-01T22:46:00Z
                </a>
            </li>
            <li class=\\"list-group-item\\">
                <a
                    href=\\"http://example.com/test/1604270730\\"
                    target=\\"_blank\\"
                    rel=\\"nofollow\\"
                >
                    Lorem ipsum 2020-11-01T22:45:30Z
                </a>
            </li>
            <li class=\\"list-group-item\\">
                <a
                    href=\\"http://example.com/test/1604270700\\"
                    target=\\"_blank\\"
                    rel=\\"nofollow\\"
                >
                    Lorem ipsum 2020-11-01T22:45:00Z
                </a>
            </li>
            <li class=\\"list-group-item\\">
                <a
                    href=\\"http://example.com/test/1604270730\\"
                    target=\\"_blank\\"
                    rel=\\"nofollow\\"
                >
                    OLD ITEM
                </a>
            </li>
            <li class=\\"list-group-item\\">
                <a
                    href=\\"http://example.com/test/16042707301111\\"
                    target=\\"_blank\\"
                    rel=\\"nofollow\\"
                >
                    NEW ITEM
                </a>
            </li>
        </ul>
    </div>
</div>
"
`;

exports[`content initial 1`] = `
"<div class=\\"row mb-5\\">
    <div class=\\"feeds col-md-10 col-lg-8 mx-auto\\"></div>
</div>
<div class=\\"row\\">
    <div class=\\"posts col-md-10 col-lg-8 mx-auto\\"></div>
</div>
"
`;
